File: python/myth_py/observer/migration.py
"""Milestone Readiness evaluation — A, B, C, D, E.

Milestone C (Gavel daemon) uses production hook-latency.ndjson only, per
Wave 3 기반점 2 (bench results are excluded from trigger evaluation).
Other milestones are Day-1 pending placeholders.
"""

from __future__ import annotations

import json
from collections.abc import Iterator
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from pathlib import Path


@dataclass
class MilestoneStatus:
    id: str  # "A" .. "E"
    title: str
    triggered: bool
    current_value: str
    threshold: str
    notes: list[str] = field(default_factory=list)


def compute_all_milestones(
    *,
    hook_latency_path: Path | None = None,
) -> list[MilestoneStatus]:
    hook_latency_path = hook_latency_path or (
        Path.home() / ".local" / "state" / "myth" / "hook-latency.ndjson"
    )
    return [
        _milestone_a(),
        _milestone_b(),
        _milestone_c(hook_latency_path),
        _milestone_d(),
        _milestone_e(),
    ]


def _milestone_a() -> MilestoneStatus:
    return MilestoneStatus(
        id="A",
        title="Assessor Tier review (3 weeks elapsed)",
        triggered=False,
        current_value="pending",
        threshold="3 weeks + tier-1 compliance < 70%",
    )


def _milestone_b() -> MilestoneStatus:
    return MilestoneStatus(
        id="B",
        title="Vector store migration",
        triggered=False,
        current_value="pending",
        threshold="N/A (design work)",
    )


def _milestone_c(latency_path: Path) -> MilestoneStatus:
    """Gavel daemon migration — production-only data, per 기반점 2."""
    if not latency_path.exists():
        return MilestoneStatus(
            id="C",
            title="The Gavel daemon migration",
            triggered=False,
            current_value="no data",
            threshold="P99 > 15ms (sustained 14d)",
            notes=[f"source: {latency_path} (production only)"],
        )

    cutoff = datetime.now(timezone.utc) - timedelta(days=14)
    latencies: list[float] = []
    for entry in _iter_ndjson(latency_path):
        if entry.get("event") != "pre_tool":
            continue
        ts = _parse_ts(entry.get("ts"))
        if ts is None or ts < cutoff:
            continue
        raw = entry.get("latency_ms")
        if isinstance(raw, (int, float)):
            latencies.append(float(raw))

    if not latencies:
        return MilestoneStatus(
            id="C",
            title="The Gavel daemon migration",
            triggered=False,
            current_value="insufficient data",
            threshold="P99 > 15ms (sustained 14d)",
            notes=[
                f"source: {latency_path} (production only)",
                "no pre_tool events in last 14 days",
            ],
        )

    latencies.sort()
    p99_index = min(int(len(latencies) * 0.99), len(latencies) - 1)
    p99 = latencies[p99_index]

    # Require minimum sample density so we do not trip on a handful of
    # startup outliers.
    min_sample = 14 * 10
    triggered = p99 > 15.0 and len(latencies) >= min_sample

    return MilestoneStatus(
        id="C",
        title="The Gavel daemon migration",
        triggered=triggered,
        current_value=f"P99: {p99:.1f}ms",
        threshold="P99 > 15ms (sustained 14d)",
        notes=[
            f"source: {latency_path} (production only)",
            f"samples: {len(latencies)} pre_tool events in 14d",
            f"min density: {min_sample}",
        ],
    )


def _milestone_d() -> MilestoneStatus:
    return MilestoneStatus(
        id="D",
        title="Semantic detection",
        triggered=False,
        current_value="pending",
        threshold="N/A (design work)",
    )


def _milestone_e() -> MilestoneStatus:
    return MilestoneStatus(
        id="E",
        title="AST validation",
        triggered=False,
        current_value="pending",
        threshold="N/A (design work)",
    )


def _iter_ndjson(path: Path) -> "Iterator[dict[str, object]]":
    with path.open(encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                yield json.loads(line)
            except json.JSONDecodeError:
                continue


def _parse_ts(value: object) -> datetime | None:
    if not isinstance(value, str):
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None

File: rust/crates/myth-ui/src/panels/migration.rs
//! Migration Readiness 패널 — 5개 Milestone A~E 상태.
//!
//! Milestone C (Gavel daemon)는 `~/.local/state/myth/hook-latency.ndjson`
//! (production source only, per Wave 3 기반점 2)에서 최근 샘플의 P99를 계산한다.
//! bench 결과는 사용하지 않는다 — bench는 구조 비용 진단용이고, 트리거
//! 판정 기준은 production warm-cache 측정이어야 한다.

use std::path::PathBuf;

use ratatui::layout::Rect;
use ratatui::style::{Modifier, Style};
use ratatui::text::{Line, Span};
use ratatui::widgets::{Block, Borders, List, ListItem};
use ratatui::Frame;
use serde::Deserialize;

use crate::theme::Theme;

#[derive(Debug, Clone)]
pub struct MilestoneStatus {
    pub id: char,
    pub title: String,
    pub triggered: bool,
    pub current_value: String,
    pub threshold: String,
    pub notes: Vec<String>,
}

pub struct MigrationPanel {
    milestones: Vec<MilestoneStatus>,
}

impl MigrationPanel {
    pub fn new() -> Self {
        let mut p = Self {
            milestones: Vec::new(),
        };
        p.refresh();
        p
    }

    pub fn milestones(&self) -> &[MilestoneStatus] {
        &self.milestones
    }

    pub fn refresh(&mut self) {
        self.milestones = vec![
            Self::milestone_a(),
            Self::milestone_b(),
            milestone_c_from_path(myth_common::hook_latency_path()),
            Self::milestone_d(),
            Self::milestone_e(),
        ];
    }

    fn milestone_a() -> MilestoneStatus {
        MilestoneStatus {
            id: 'A',
            title: "Assessor Tier review (3w elapsed)".into(),
            triggered: false,
            current_value: "pending".into(),
            threshold: "3 weeks".into(),
            notes: vec![],
        }
    }

    fn milestone_b() -> MilestoneStatus {
        MilestoneStatus {
            id: 'B',
            title: "Vector store migration".into(),
            triggered: false,
            current_value: "pending".into(),
            threshold: "N/A".into(),
            notes: vec![],
        }
    }

    fn milestone_d() -> MilestoneStatus {
        MilestoneStatus {
            id: 'D',
            title: "Semantic detection".into(),
            triggered: false,
            current_value: "pending".into(),
            threshold: "N/A".into(),
            notes: vec![],
        }
    }

    fn milestone_e() -> MilestoneStatus {
        MilestoneStatus {
            id: 'E',
            title: "AST validation".into(),
            triggered: false,
            current_value: "pending".into(),
            threshold: "N/A".into(),
            notes: vec![],
        }
    }

    pub fn render(&self, frame: &mut Frame, area: Rect, focused: bool, theme: &Theme) {
        let block = Block::default()
            .title(" Migration Readiness ")
            .borders(Borders::ALL)
            .border_style(theme.border_style(focused));

        let items: Vec<ListItem> = self
            .milestones
            .iter()
            .map(|m| {
                let marker = if m.triggered { "[x]" } else { "[ ]" };
                let color = if m.triggered { theme.warn } else { theme.dim };
                ListItem::new(Line::from(vec![
                    Span::styled(
                        format!("{} [{}] ", m.id, marker),
                        Style::default().fg(color).add_modifier(Modifier::BOLD),
                    ),
                    Span::styled(m.title.clone(), Style::default().fg(theme.fg)),
                    Span::styled(
                        format!("  {} / {}", m.current_value, m.threshold),
                        Style::default().fg(theme.accent),
                    ),
                ]))
            })
            .collect();
        let list = List::new(items).block(block);
        frame.render_widget(list, area);
    }
}

impl Default for MigrationPanel {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Deserialize)]
struct HookLatencyEntry {
    #[serde(default)]
    latency_ms: f64,
}

/// Milestone C = Gavel daemon trigger. 최근 14일 P99 > 15ms 기준.
/// 파일이 없거나 샘플이 부족하면 untriggered + "insufficient data".
pub fn milestone_c_from_path(path: PathBuf) -> MilestoneStatus {
    let Ok(content) = std::fs::read_to_string(&path) else {
        return MilestoneStatus {
            id: 'C',
            title: "Gavel daemon migration".into(),
            triggered: false,
            current_value: "no data".into(),
            threshold: "P99 > 15ms".into(),
            notes: vec![format!("source: {}", path.display())],
        };
    };

    let mut samples: Vec<f64> = Vec::new();
    for line in content.lines() {
        if let Ok(e) = serde_json::from_str::<HookLatencyEntry>(line) {
            samples.push(e.latency_ms);
        }
    }

    if samples.is_empty() {
        return MilestoneStatus {
            id: 'C',
            title: "Gavel daemon migration".into(),
            triggered: false,
            current_value: "insufficient data".into(),
            threshold: "P99 > 15ms".into(),
            notes: vec![format!("source: {}", path.display())],
        };
    }

    let p99 = percentile(&mut samples, 0.99);
    let triggered = p99 > 15.0;
    MilestoneStatus {
        id: 'C',
        title: "Gavel daemon migration".into(),
        triggered,
        current_value: format!("P99 {p99:.1}ms"),
        threshold: "P99 > 15ms".into(),
        notes: vec![
            format!("source: {} (production only)", path.display()),
            format!("samples: {}", samples.len()),
        ],
    }
}

fn percentile(samples: &mut [f64], q: f64) -> f64 {
    samples.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    if samples.is_empty() {
        return 0.0;
    }
    let idx = ((samples.len() as f64 - 1.0) * q).round() as usize;
    samples[idx.min(samples.len() - 1)]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn milestone_c_missing_file() {
        let ms = milestone_c_from_path(PathBuf::from("/nonexistent/hook-latency.ndjson"));
        assert_eq!(ms.id, 'C');
        assert!(!ms.triggered);
        assert_eq!(ms.current_value, "no data");
    }

    #[test]
    fn milestone_c_empty_file() {
        let tmp = tempfile::tempdir().unwrap();
        let p = tmp.path().join("hook.ndjson");
        std::fs::write(&p, "").unwrap();
        let ms = milestone_c_from_path(p);
        assert!(!ms.triggered);
        assert_eq!(ms.current_value, "insufficient data");
    }

    #[test]
    fn milestone_c_low_latency_not_triggered() {
        let tmp = tempfile::tempdir().unwrap();
        let p = tmp.path().join("hook.ndjson");
        let content: String = (1..=100)
            .map(|i| format!(r#"{{"latency_ms":{}}}"#, i as f64 * 0.1))
            .collect::<Vec<_>>()
            .join("\n");
        std::fs::write(&p, content).unwrap();
        let ms = milestone_c_from_path(p);
        assert!(!ms.triggered, "P99 {} shouldn't exceed 15ms", ms.current_value);
    }

    #[test]
    fn milestone_c_high_latency_triggered() {
        let tmp = tempfile::tempdir().unwrap();
        let p = tmp.path().join("hook.ndjson");
        let content: String = (1..=100)
            .map(|i| format!(r#"{{"latency_ms":{}}}"#, i as f64))
            .collect::<Vec<_>>()
            .join("\n");
        std::fs::write(&p, content).unwrap();
        let ms = milestone_c_from_path(p);
        assert!(ms.triggered, "P99 {} should exceed 15ms", ms.current_value);
    }

    #[test]
    fn percentile_basic() {
        let mut s = vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0];
        let p50 = percentile(&mut s, 0.5);
        assert!((5.0..=6.0).contains(&p50));
    }

    #[test]
    fn refresh_populates_five() {
        let p = MigrationPanel::new();
        assert_eq!(p.milestones().len(), 5);
        let ids: Vec<char> = p.milestones().iter().map(|m| m.id).collect();
        assert_eq!(ids, vec!['A', 'B', 'C', 'D', 'E']);
    }
}
